sql,externalconn: add ALTER EXTERNAL CONNECTION

edwardguo-crl opened the pull request:
## Summary

Add AST parser support for the new SQL command `ALTER EXTERNAL CONNECTION`, allowing users to modify existing external connections.

Fixes #98610

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146829)
<!-- Reviewable:end -->

msbutler reviewed (commented):
nice start!

edwardguo-crl commented on pkg/sql/parser/sql.y:
```diff
@@ -3830,6 +3834,26 @@ create_external_connection_stmt:
+// %Help: ALTER EXTERNAL CONNECTION - alter a existing external connection
+// %Category: Misc
+// %Text:
+// ALTER EXTERNAL CONNECTION [IF NOT EXISTS] <name> AS <endpoint>
```
Gotcha, I will update the here and the implmentation.

edwardguo-crl commented on pkg/cloud/externalconn/record.go:
```diff
@@ -110,6 +110,22 @@ func LoadExternalConnection(
 
+// ExternalConnectionExist loads an mutableExternalConnection from system table. Allow owner to modifiy the URI.
+func ExternalConnectionExist(ctx context.Context, name string, ownerID oid.Oid, txn isql.Txn) error {
+	row, _, err := txn.QueryRowExWithCols(ctx, "lookup-schedule", txn.KV(),
```
Hey @msbutler,

I am not sure if I should use `lookup-schedule` or something else?

edwardguo-crl commented on pkg/cloud/externalconn/record.go:
```diff
@@ -354,6 +370,44 @@ func (e *MutableExternalConnection) Create(ctx context.Context, txn isql.Txn) er
+	// CREATE EXTERNAL CONNECTION is only allowed for users with the
+	// `EXTERNALCONNECTION` system privilege. We run the query as `node`
+	// since the user might not have `INSERT` on the system table.
+	createQuery := "UPSERT INTO system.external_connections (%s) VALUES (%s) RETURNING connection_name"
```
Should I use the standard PostgreSQL syntax here—e.g., `INSERT … ON CONFLICT DO UPDATE`? Can I assume the internal SQL planner still uses CRDB?

CC: @msbutler

edwardguo-crl commented on pkg/cloud/externalconn/record.go:
```diff
@@ -354,6 +370,44 @@ func (e *MutableExternalConnection) Create(ctx context.Context, txn isql.Txn) er
+		return err
+	}
+
+	// CREATE EXTERNAL CONNECTION is only allowed for users with the
```
oh no, I will update the comment here.

edwardguo-crl commented on pkg/cloud/externalconn/record.go:
```diff
@@ -354,6 +370,44 @@ func (e *MutableExternalConnection) Create(ctx context.Context, txn isql.Txn) er
+	if err != nil {
+		// If we see a UniqueViolation it means that we are attempting to create an
+		// External Connection with a `connection_name` that already exists.
+		if pgerror.GetPGCode(err) == pgcode.UniqueViolation {
```
I see that `CREATE EXTERNAL CONNECTION` already enforces uniqueness. Can we skip that check for `ALTER EXTERNAL CONNECTION`?

msbutler commented:
@edwardguo-crl looks like the build failed here. As i mentioned before, could you split this PR in 2: one for the syntax change, and then a follow up PR with the logic in the externalconn pkg?

edwardguo-crl committed:
sql,externalconn: add ALTER EXTERNAL CONNECTION Parser

Add AST parser support for the new SQL command ALTER EXTERNAL
CONNECTION, allowing users to modify existing external connections.

Fixes #98610

msbutler reviewed (approved):
Nice!
